File: best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int profit = 0;
        int maxPeak = prices[prices.length-1];

        for(int i = prices.length-1; i >= 0; i--) {
            maxPeak = Math.max(maxPeak,prices[i]);
            profit = Math.max(profit, maxPeak - prices[i]);
        }

        return profit;
    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0
        lowestSoFar = float('inf')
        for i in range(len(prices)):
            lowestSoFar = min(lowestSoFar, prices[i])
            profit = max(profit, prices[i] - lowestSoFar)
        return profit
          